trin-core/src/types/validation.rs

```diff
@@ -0,0 +1,101 @@
+use anyhow::anyhow;
+use async_trait::async_trait;
+use serde_json::{json, Value};
+
+use crate::jsonrpc::service::dispatch_infura_request;
+use crate::jsonrpc::types::{HistoryJsonRpcRequest, JsonRequest, Params};
+use crate::portalnet::types::content_key::IdentityContentKey;
+use crate::portalnet::types::messages::ByteList;
+
+// This struct is responsible for dispatching cross-overlay-network requests
+// for data to perform validation. Currently, it just proxies these requests
+// on to infura.
+#[derive(Clone)]
+pub struct HeaderOracle {
+    pub infura_url: String,
+    // We could simply store the main portal jsonrpc tx channel here, rather than each
+    // individual channel. But my sense is that this will be more useful in terms of
+    // determining which subnetworks are actually available.
+    pub history_jsonrpc_tx: Option<tokio::sync::mpsc::UnboundedSender<HistoryJsonRpcRequest>>,
+    pub header_gossip_jsonrpc_tx: Option<bool>,
+    pub block_indices_jsonrpc_tx: Option<bool>,
+    pub state_jsonrpc_tx: Option<bool>,
```

Comment:
I agree with having these individually instead of the main portal jsonrpc tx channel for the reason you said, but down the line maybe we want to wrap these `jsonrpc_tx`s up into a wrapper struct to avoid duplicating the same four `jsonrpc_tx` members in the `BlockBodyOracle` and `BlockReceiptsOracle`. Though that's only if all oracles need access to all networks. If not, I think the clearest thing would be to have each oracle have access to the minimum number of networks necessary.

Reply:
There may be some confusion in how I explained this, but the idea was to only have a single "Oracle" - the object that has access to the various tx channels. This will be shared (via `Arc<RwLock>`) between the different "Validators" (eg. `BlockBodyValidator` and `ReceiptsValidator`) which will contain all the type-specific logic for validation. Iiuc - then the `HeaderOracle` is already behaving like the wrapper struct you're describing here?

Reply:
Hmm, I'm not seeing why there would be a `BlockBodyValidator` and `ReceiptsValidator` (instead of `BlockBodyOracle` and `ReceiptsOracle`) when currently there's `ChainHistoryValidator` and `StateValidator`. Seems like it's currently set up so that each network has one validator, and each data type has an oracle, but you're saying there will be a single oracle (in which case, `HeaderOracle` sounds like it's overly specific to headers) and a validator per data type.

Reply:
Ahh, my bad I got mixed up. You're right. The `ChainHistoryValidator` will do all validation for chain history network types (headers, bodies, and receipts). But, yeah the name `HeaderOracle` isn't necessarily an oracle for just the `BlockHeader` type. `HeaderOracle` refers more to the idea of an oracle that can answer questions like `is_hash_canonical`, `get_hash_at_height`, `get_latest_header` (etc). Which might be used to validate state network data for example. This is one reason why I considered the name `ValidationOracle` - which seems like a more generic name, but since we've used `HeaderOracle` in discussion I went that way. However, we don't need a `BlockBodyOracle` and `ReceiptsOracle` - we can simply use the `HeaderOracle` to validate those data types.